### DISCUSSION TITLE
sendMessage freeze if I try send more than 20 message / per minute for different users

### DISCUSSION THREAD

**Discussion by sautchenkov** [rubenlagus/TelegramBots#1344] [General]
Hello! I try to send a message to all bot users. In the documentation, I see the only 30/sec limit. But when I try to send, if I send more than 20/min sendMessage freeze for 30 sec. And I can send more than 20/per minute to separate users. Does anyone get this limit?

**Comment by gekoramy** [comment]
https://github.com/rubenlagus/TelegramBots/issues/238

I'm aware of these limits:
- max 1 message every 90ms to different chats
- max 1 message every second to the same chat
- max 20 messages every minute to the same chat

If you don't use inline buttons on your messages and respect these limits your bot shouldn't experience any block

**Comment by sautchenkov** [comment]
@gekoramy Thanks for the answer.  I using inline buttons in messages. How can this affect?

**Comment by gekoramy** [comment]
My bot sends messages with inline buttons and respects the limits I wrote before.
Unfortunately, evil users are still able to do a DoS attack by simply tapping repeatedly those buttons.
The good things about my bot are that:
- completing the DOS attack takes more than 1 minute
- normal users will hardly ever cause an "accidental" DoS attack.

In my opinion, the best way to prevent "real" DoS attack is to set cache time to x seconds on inline buttons.
In this way, users who tap repeatedly the buttons will send just a request every x seconds.
Obviously, this solution is possible only if you use inline buttons for certain actions.
